src/Buttons/ButtonDropdown.tsx

```diff
@@ -0,0 +1,166 @@
+   */
+  options: Array<DropdownOptionProps>;
+  /** Specifies the alt text for screen readers */
+  alt: string;
+  /** Specifies the ref of the element */
+  forwardRef?: Ref<HTMLButtonElement>;
+}
+
+const StyledButtonDropdown = styled.button`
+  background: ${fromTheme('color', 'background', 'medium')};
+  border: none;
+  cursor: pointer;
+  display: inline-block;
+  font-size: ${fromTheme('font', 'body', 'size')};
+  font-weight: ${fromTheme('font', 'body', 'weight')};
+  padding: ${({ theme }) => (`${theme.ws.xsmall} ${theme.ws.small}`)};
+  &:hover {
+    background: ${fromTheme('color', 'background', 'dark')};
+    border: none;
+    color: ${fromTheme('color', 'text', 'base')};
+  }
+`;
+
+const StyledMenu = styled.div<MenuProps>`
+  background: ${fromTheme('color', 'background', 'light')};
+  border: ${fromTheme('border', 'hairline')};
+  min-height: 3em;
+  min-width: 6em;
+  position: absolute;
+  transform: translate3d(0em, 0.1em, 0em);
+  z-index: 100;
+`;
+
+const StyledMenuList = styled.ul`
+  list-style-type: none;
+`;
+
+const StyledMenuListItem = styled.li`
+  padding: .25em 1em;
+  &:hover {
+    background: ${fromTheme('color', 'background', 'medium')};
+    cursor: pointer;
+  }
+`;
+
+const StyledMenuButton = styled.button`
+  background: transparent;
+  border: none;
+  &:hover {
+    background: ${fromTheme('color', 'background', 'medium')};
+    cursor: pointer;
+  }
+`;
+
+/**
+ * A component that allows users to specify a Font Awesome icon to be displayed
+ * in the button and options within the dropdown. When an option is clicked, the
+ * value that was selected is passed to the onChange function and the dropdown
+ * menu collapses.
+ */
+const ButtonDropdown: FunctionComponent<ButtonDropdownProps> = (props)
+: ReactElement => {
+  const {
+    id,
+    children,
+    onChange,
+    forwardRef,
+    alt,
+    options,
+  } = props;
+
+  const [isMenuVisible, setMenuVisible] = useState(false);
+  const theme = useContext(ThemeContext);
+
+  // When the menu is open and the user clicks on the document, it minimizes the menu
+  const handleDocumentBodyClick = useCallback(() => {
+    setMenuVisible(false);
+  }, [setMenuVisible]);
+
+  // Removes the event listener after the click event occurs
+  useEffect(() => {
+    if (isMenuVisible) {
+      document.addEventListener('click', handleDocumentBodyClick);
+    } else {
+      document.removeEventListener('click', handleDocumentBodyClick);
+    }
+  }, [isMenuVisible, handleDocumentBodyClick]);
+
+  return (
+    <>
+      <StyledButtonDropdown
+        id={id}
+        onClick={() => {
+          setMenuVisible(!isMenuVisible);
+        }}
+        theme={theme}
+        ref={forwardRef}
+        aria-label={alt}
+      >
+        { children }
+      </StyledButtonDropdown>
+      {isMenuVisible && (
+        <>
+          <StyledMenu isMenuVisible={isMenuVisible}>
+            <StyledMenuList>
+              {options.map((option) => (
```

Comment:
Thinking about this some more, I'm starting to feel like it would make more sense to change `ButtonDropdown` to `ButtonDropdownMenu` and separate export a `ButtonDropdownMenuItem` component. Then we could let the consuming application compose the menu however it wants by making the contents of the `StyledMenuList` be the children of `ButtonDropdownMenu`, similar to how the `TabList` and `TabListItem` work. (In that case, we'd want to add a `label` prop to that `ButtonDropdownMenu` to set the content of the `Button`). My thinking here is:

1.) We wouldn't need to be as limited in the shape of `options` and how the `onChange` handler gets called, since that would be done as part of the implementation of `ButtonDropdownMenuItem`. For instance, in the course-planner, we could have `startTime` and `endTime` props that could be handled separately. So something like:

```ts
// in course planner

<ButtonDropdownMenu label={<FontAwesomeIcon />}>
  {[
    { 
      label: '1:00 PM - 2:00 PM',
      startTime: '13:00:00',
      endTime: '14:00:00',
    }, 
    ...rest,
  ].map(({label, startTime, endTime}) => (
    <ButtonDropdownMenuItem
      onClick={() => {
        // state setters from parent
        setStartTime(startTime);
        setEndTime(endTime);
      })
    >
      {label}
    <ButtonDropdownMenuItem>
  ))}
</ButtonDropdownMenu>

2.) We could potentially add non-button items to the list, e.g. headings or links

I may be trying to over-optimize, but that feels like a more react-y way of doing things. What do you think?

Reply:
Thanks for outlining how the code would look in course planner - that's a huge help for me! I think the changes were addressed in commit 15db8d9.